### DISCUSSION TITLE
Is there a way to override Visual Studio Keybindings in extension?

### DISCUSSION THREAD

**Discussion by aglasencnik** [VsixCommunity/Community.VisualStudio.Toolkit#459] [Q&A]
I am wondering if there is a way of overriding default Visual Studio Keybinding behaviours in my extension?

Basically I am developing an extension and I have quite a lot of keybindings set. But because of safety all of them have a "prefix" keys which come before the actual keys eg. Ctrl + Shift + Alt + E (prefix) then Ctrl + D.

I would want to remove those "prefix" keys, so I would just have the actual keys like Ctrl + D, I thought if I could check if those were pressed or something and if my action/command did some stuff then the default behaviour wouldn't go through, and if my extensions action doesn't go through it must do the default behaviour.

Is there any way to achieve this?

**Comment by madskristensen** [comment]
I haven't found a way to change existing keybindings from an extension, unfortunately.

**Reply by aglasencnik** [reply]
Hi, thanks for your reply... I wouldn't change them, I would just intercept the call...
Dont you have something similar in your ZenCoding extension? You listen for the Tab key...

**Comment by madskristensen** [comment]
Ahh, yes you can intercept either an existing command and take it over, or you can listen to a keypress

**Reply by aglasencnik** [reply]
Ok cool so it is possible? Would you mind sharing an example of how to do something like that? I think listening for keypresses would be a better option for me as I in intercepting existing commands I wouldn't really know which to intercept if you know what I mean or what do you think would be better in my case? In listening for the keypresses can I listen and get a combination (eg. Ctrl + D)? I would really appreciate it if you'd provide me with anything to help me achieve this, as this is my first extension and I think it could be quite useful to a lot of people.

**Reply by madskristensen** [reply]
If you have an editor extension, then you can listen for key presses, but if it should be global, then you need to take over a command. 

To take over a command, use this method:

```csharp
VS.Commands.InterceptAsync()
```

To listen for keystrokes in the editor, see this https://github.com/madskristensen/Surrounder/blob/e2aefe27cf806f7a3a37568b57a475579768094b/src/InsertTextCommand.cs#L16

**Reply by aglasencnik** [reply]
Ok, should I have the keybindings set anyways in the .vsct file?
Also in the keystrokes example I dont understand where you check it.... I looked through the code

**Comment by madskristensen** [comment]
For the keystrokes, you check it like this: https://github.com/madskristensen/Surrounder/blob/e2aefe27cf806f7a3a37568b57a475579768094b/src/InsertTextCommand.cs#L37

No, you don't need .vsct to intercept a command

**Reply by aglasencnik** [reply]
Ok this is much better now, but still there is an issue on how could I see if eg. Ctrl + D or Ctrl + Up arrow,... were pressed? Because those command args seem to just get the single character or Tab,....

**Reply by aglasencnik** [reply]
Also I have been trying the way of overriding those commands with: IOleCommandTarget? But I can't seem to get it to work.... Basically it is never getting hit (even just the method).... Also for that I have made the keybindings and Id's for the command but still beside the button I don't see the key combination like I do where I have that prefix from before...

**Reply by aglasencnik** [reply]
Also it is strange because for me even ICommandHandler doesn't seem to work... Here is my package file code:
```csharp
global using Community.VisualStudio.Toolkit;
global using Microsoft.VisualStudio.Shell;
global using System;
global using Task = System.Threading.Tasks.Task;
using Microsoft.VisualStudio.Shell.Interop;
using System.Runtime.InteropServices;
using System.Threading;

namespace AdvancedKeyBindingsTest;

[PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading = true)]
[InstalledProductRegistration(Vsix.Name, Vsix.Description, Vsix.Version)]
[ProvideMenuResource("Menus.ctmenu", 1)]
[Guid(PackageGuids.InsertGuidString)]
[ProvideAutoLoad(UIContextGuids.NoSolution, PackageAutoLoadFlags.BackgroundLoad)]
[ProvideAutoLoad(UIContextGuids.SolutionExists, PackageAutoLoadFlags.BackgroundLoad)]
public sealed class AdvancedKeyBindingsTestPackage : ToolkitPackage
{
    protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
    {
        await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
        await this.RegisterCommandsAsync();
    }
}
```
And the HandleTabCommand:
```csharp
using Microsoft.VisualStudio.Commanding;
using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Utilities;
using System.ComponentModel.Composition;

namespace AdvancedKeyBindingsTest.Commands;

[Export(typeof(ICommandHandler))]
[Name(nameof(HandleTabCommand))]
[ContentType("text")]
[TextViewRole(PredefinedTextViewRoles.PrimaryDocument)]
internal class HandleTabCommand : ICommandHandler<TabKeyCommandArgs>
{
    public string DisplayName => nameof(HandleTabCommand);

    public bool ExecuteCommand(TabKeyCommandArgs args, CommandExecutionContext executionContext)
    {
        return true; 
    }

    public CommandState GetCommandState(TabKeyCommandArgs args)
    {
        return CommandState.Available;
    }
}
```

**Reply by RobertvanderHulst** [reply]
I think your commands have to inherit from `BaseCommand<T>`.
See https://github.com/X-Sharp/XSharpPublic/blob/main/src/VisualStudio/ProjectPackage/Commands/CommandEditProjectFile.cs for an example

**Reply by aglasencnik** [reply]
@RobertvanderHulst, unfortunately that isn't the case here... As I still can't understand how when I tested Mads' ZenCoding extension in debugging the key handling worked (for Tab... I would need Ctrl and keys), but when I created almost the same in mine it didn't work....

**Reply by RobertvanderHulst** [reply]
@aglasencnik The way that I understand it, the RegisterCommandsAsync (extension) method  of the AsyncPackage class checks for all types that inherit from `BaseCommand<T>` and call the initializeAsync method to create an object of these types.

**Reply by aglasencnik** [reply]
@RobertvanderHulst yes but my command is not the issue here... I want to implement keybinding overrides then I will do stuff with the commands

**Reply by aglasencnik** [reply]
Ok so update... I found out that I have to add an asset the project to MefComponent in manifest file... Now the handler for Tab works... BUT I still somehow need to find a way to work with key combinations like Ctrl + D, Ctrl + Shift + Up arrow,....

**Reply by aglasencnik** [reply]
Another update.... I have an idea on how to process and check if the keys are pressed (example code is below):
```csharp
using Microsoft.VisualStudio.Commanding;
using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Utilities;
using System.ComponentModel.Composition;

namespace AdvancedKeyBindingsTest;

[Export(typeof(ICommandHandler))]
[Name(nameof(HandleKeyPress))]
[ContentType(ContentTypes.Text)]
[TextViewRole(PredefinedTextViewRoles.PrimaryDocument)]
internal class HandleKeyPress : ICommandHandler<TypeCharCommandArgs>
{
    public string DisplayName => Vsix.Name;

    public bool ExecuteCommand(TypeCharCommandArgs args, CommandExecutionContext executionContext)
    {
        var modifiers = System.Windows.Input.Keyboard.Modifiers;
        var isCtrlPressed = (modifiers & System.Windows.Input.ModifierKeys.Control) != 0;
        var isShiftPressed = (modifiers & System.Windows.Input.ModifierKeys.Shift) != 0;

        if (isCtrlPressed && isShiftPressed && args.TypedChar == 'I')
        {
            return true;
        }

        return false;
    }

    public CommandState GetCommandState(TypeCharCommandArgs args)
    {
        return CommandState.Available;
    }
}
```
But there is now another problem... It doesn't override the other commands which have Ctrl key in them.... So basically everything with Ctrl doesn't work... Any ideas on what I can do?
